sqlmesh/core/engine_adapter/snowflake.py

```diff
@@ -103,15 +103,44 @@ def _fetch_native_df(
             columns = self.cursor._result_set.batches[0].column_names
             return pd.DataFrame([dict(zip(columns, row)) for row in rows])
 
-    def _get_data_objects(self, schema_name: SchemaName) -> t.List[DataObject]:
+    def _get_data_objects(
+        self, schema_name: SchemaName, object_names: t.Optional[t.Set[str]] = None
+    ) -> t.List[DataObject]:
         """
         Returns all the data objects that exist in the given schema and optionally catalog.
         """
         from snowflake.connector.errors import ProgrammingError
 
-        sql = f'SHOW TERSE OBJECTS IN "{to_schema(schema_name).sql(dialect=self.dialect)}"'
+        schema = to_schema(schema_name)
+        catalog_name = schema.catalog or self.get_current_catalog()
+        schema_sql = schema.sql(dialect=self.dialect)
+        query = (
+            exp.select(
+                exp.column("TABLE_CATALOG").as_("catalog"),
+                exp.column("TABLE_NAME").as_("name"),
+                exp.column("TABLE_SCHEMA").as_("schema_name"),
+                exp.case()
+                .when(exp.column("TABLE_TYPE").eq("BASE TABLE"), exp.Literal.string("TABLE"))
+                .when(exp.column("TABLE_TYPE").eq("TEMPORARY TABLE"), exp.Literal.string("TABLE"))
+                .when(exp.column("TABLE_TYPE").eq("EXTERNAL TABLE"), exp.Literal.string("TABLE"))
+                .when(exp.column("TABLE_TYPE").eq("EVENT TABLE"), exp.Literal.string("TABLE"))
+                .when(exp.column("TABLE_TYPE").eq("VIEW"), exp.Literal.string("VIEW"))
+                .when(
+                    exp.column("TABLE_TABLE").eq("MATERIALIZED VIEW"),
+                    exp.Literal.string("MATERIALIZED_VIEW"),
+                )
+                .else_(exp.column("TABLE_TYPE"))
+                .as_("type"),
+            )
+            .from_(exp.table_("TABLES", db="INFORMATION_SCHEMA", catalog=catalog_name))
+            .where(exp.column("TABLE_SCHEMA").eq(schema.db))
+        )
+        if object_names:
+            # FIXME: Find a way to properly normalize object names.
```

Comment:
arent names already normalized? where do they come from?

Reply:
Actually you might be right